### PR TITLE
Change geometric transform inverse to property

### PR DESCRIPTION
**Pull request by aeisenbarth**
## Description

This PR changes the signature of `inverse` for all transform classes to a property returning a new instance of the same type with the inverse transform. When called with coordinates, the new transform replicates the behavior of the former `inverse(coords)` method.

This enables:
- accessing parameters of an inverse, e.g. `tform.inverse.params`
- composing transforms with the inverse first, e.g. `tform1.inverse + tform2`
- inverting an inverse, e.g. `tform.inverse.inverse`

Closes issue #6921.

## Checklist

- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ ] Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

File: skimage/transform/tests/test_geometric.py [hunk 1]
```diff
@@ -544,6 +544,39 @@ def test_union_differing_types():
...
+        EuclideanTransform(rotation=0.9, translation=(5, 5)),
+        SimilarityTransform(scale=0.1, rotation=0.9),
+        EssentialMatrixTransform(
+            rotation=np.eye(3), translation=(1 / np.sqrt(2), 1 / np.sqrt(2), 0)
+        ),
+        FundamentalMatrixTransform(
+            matrix=EssentialMatrixTransform(
+                rotation=np.eye(3), translation=(1 / np.sqrt(2), 1 / np.sqrt(2), 0)
+            ).params
+        ),
```
Comment (aeisenbarth) [inline skimage/transform/tests/test_geometric.py:561]: Not sure, whether the style of initializing test data within the function or pytest parametrization is preferred. A downside of this is that the transforms are created at load time.

File: skimage/transform/tests/test_geometric.py [hunk 2]
```diff
@@ -544,6 +544,39 @@ def test_union_differing_types():
...
+        ((t := PiecewiseAffineTransform()).estimate(SRC, DST) and t),
```
Comment (aeisenbarth) [inline skimage/transform/tests/test_geometric.py:562]: This is a slightly unobvious way to initialize the parameters of PiecewiseAffineTransform because the constructor does not accept arguments and `estimate` does not return the transform. This corresponds to:
```python
tform = PiecewiseAffineTransform()
tform.estimate(SRC, DST)
tform
```

File: skimage/transform/tests/test_geometric.py [hunk 3]
```diff
@@ -544,6 +544,39 @@ def test_union_differing_types():
...
+)
+def test_inverse(tform):
+    assert isinstance(tform.inverse, type(tform))
+    assert_almost_equal(tform.inverse.inverse(SRC), tform(SRC))
+    if not isinstance(
+        tform,
+        (
+            EssentialMatrixTransform,
+            FundamentalMatrixTransform,
+            PiecewiseAffineTransform,
```
Comment (aeisenbarth) [inline skimage/transform/tests/test_geometric.py:573]: This transforms have no implementation for `__add__`.

**Comment by aeisenbarth** [comment]
I couldn't find release_dev.rst.

File: skimage/transform/tests/test_geometric.py (hunk 1 above)
Comment (lagru) [reply skimage/transform/tests/test_geometric.py:561]: It's a bit weird, but the entire test suite for this module isn't well structured (e.g. no grouping of tests by class). I'm not really happy but the same goes for alternatives I can think of.

**Review by lagru — COMMENTED** [review]
@aeisenbarth, thanks so much! This is looking great. I don't have any major complains but would like some clarification in a few places and added a few suggestions.

**Comment by lagru** [comment]
> I couldn't find release_dev.rst.

It's in [doc/source/release_notes/release_dev.rst](https://github.com/scikit-image/scikit-image/blob/main/doc/source/release_notes/release_dev.rst). Though, to be honest we haven't really been diligent about summarizing our PRs there. :see_no_evil: I hope that we can improve on this in the future with a better solution...

File: skimage/transform/_geometric.py [hunk 4]
```diff
@@ -293,23 +282,10 @@ def __call__(self, coords):
...
-            Epipolar lines in the source image.
-
-        """
-        coords = np.asarray(coords)
-        coords_homogeneous = np.column_stack([coords, np.ones(coords.shape[0])])
-        return coords_homogeneous @ self.params
+    @property
+    def inverse(self):
+        """Return a transform object representing the inverse."""
+        return type(self)(matrix=self.params.T)
```
[the remark this answers is not in the document]
Comment (aeisenbarth) [reply skimage/transform/_geometric.py:288]: I'm not so familiar with epipolar geometry, but intuitively I would understand the term "inverse" on all transform classes as something capable of the backward transform from DST to SRC. I see this independent from whether the parameters are polynomial coefficients or a fundamental matrix; and whether the backwards transformation is actually performed by a transposed matrix instead of an inverse matrix.

Other opinions would be helpful!

File: skimage/transform/_geometric.py [hunk 5]
```diff
@@ -810,10 +775,6 @@ def __add__(self, other):
             else:
                 tform = ProjectiveTransform
             return tform(other.params @ self.params)
-        elif (hasattr(other, '__name__')
-                and other.__name__ == 'inverse'
-                and hasattr(get_bound_method_class(other), '_inv_matrix')):
-            return ProjectiveTransform(other.__self__._inv_matrix @ self.params)
```
[the remark this answers is not in the document]
Comment (aeisenbarth) [reply skimage/transform/_geometric.py:816]: I considered this as obsolete.

The `elif` would check whether the passed `other` is not a transform instance but the `inverse` method object. Then it would get from the receiver (transform object) the inverse matrix attribute.

It would still be useful if someone was using multiple versions of scikit-image in the same code (so an transform inverse method could be added to a new transform object) or if we want to keep duck-typing-like compatibility to non-scikit-image transforms. I can revert it back.

File: skimage/transform/tests/test_geometric.py [hunk 6]
```diff
@@ -544,6 +544,39 @@ def test_union_differing_types():
...
+            matrix=EssentialMatrixTransform(
+                rotation=np.eye(3), translation=(1 / np.sqrt(2), 1 / np.sqrt(2), 0)
+            ).params
+        ),
+        ((t := PiecewiseAffineTransform()).estimate(SRC, DST) and t),
+    ],
+)
+def test_inverse(tform):
+    assert isinstance(tform.inverse, type(tform))
+    assert_almost_equal(tform.inverse.inverse(SRC), tform(SRC))
```
[the remark this answers is not in the document]
Comment (aeisenbarth) [reply skimage/transform/tests/test_geometric.py:567]: I actually had it more explicit like this, but it only works for ProjectiveTransform since the others have other attribute names. I resorted to the more generic way of applying the transform to points.

I could add a conditional check for `params`, or split the test into one transforms that support `params` and one for the rest.

File: skimage/transform/_geometric.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lagru) [reply skimage/transform/_geometric.py:816]: Ah, true. It checks whether the name of `other` is `inverse`; I actually read that as checking for having the attribute `inverse`. I'm onboard with removing it then. :+1: I don't think the mentioned remaining cases were ever intended to be part of the API.

File: skimage/transform/_geometric.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lagru) [reply skimage/transform/_geometric.py:288]: Reading your response and thinking about it again, inverse *is* a pretty accepted term describing the opposite operation of a transformation (e.g. `ifft`, `iradon`). Let's keep it like this.

File: skimage/transform/tests/test_geometric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lagru) [reply skimage/transform/tests/test_geometric.py:567]: Huh, I tested this locally. The only one missing the `params` attribute is the `PiecewiseAffineTransform`? Though for the test's to pass the more relaxed `assert_almost_equal` is required rather than `assert_array_equal`. Maybe something like this
```suggestion
    try:
        assert_almost_equal(tform.inverse.inverse.params, tform.params)
    except AttributeError:
        assert isinstance(tform, PiecewiseAffineTransform)
    assert_almost_equal(tform.inverse.inverse(SRC), tform(SRC))
```
Though not really an improvement in readability. I'll leave it up to you if and how you want to include it. :)

**Review by lagru — APPROVED** [review]
Thank you for the very quick and high quality contribution @aeisenbarth! Ready to go in my opinion. :)

**Comment by stefanv** [comment]
> > I couldn't find release_dev.rst.
> 
> It's in [doc/source/release_notes/release_dev.rst](https://github.com/scikit-image/scikit-image/blob/main/doc/source/release_notes/release_dev.rst). Though, to be honest we haven't really been diligent about summarizing our PRs there. Rather we went through every merged PR while preparing the release. see_no_evil I hope that we can improve on this in the future with a better solution...

The solution is to make smaller releases, list all PRs, and do minimal editing of release notes. We should update our contributor guidelines to remove the release notes requirement.

File: skimage/transform/_geometric.py (hunk 4 above)
Comment (stefanv) [inline skimage/transform/_geometric.py:288]: Why is the inverse given by the transpose?

File: skimage/transform/_geometric.py [hunk 7]
```diff
@@ -1148,40 +1109,15 @@ def __call__(self, coords):
...
-            # all coordinates within triangle
-            index_mask = simplex == index
-
-            out[index_mask, :] = affine(coords[index_mask, :])
-
-        return out
+    @property
+    def inverse(self):
+        """Return a transform object representing the inverse."""
+        tform = type(self)()
```
Comment (stefanv) [inline skimage/transform/_geometric.py:1115]: Maybe `PiecewiseAffineTransform()`?

File: skimage/transform/tests/test_geometric.py (hunk 1 above)
Comment (stefanv) [reply skimage/transform/tests/test_geometric.py:561]: Yeah, I wouldn't worry too much about it.

**Comment by stefanv** [comment]
Great contribution, thank you!

**Comment by aeisenbarth** [comment]
Oh, and something I noticed:
`PolynomialTransform.inverse` raises an `Exception`.
https://github.com/scikit-image/scikit-image/blob/7453d93e8a174d0cf55eebf742d5ae58d9b5cf4a/skimage/transform/_geometric.py#L1668-L1673

1. Shouldn't it rather raise a specific exception class or `NotImplementedError`?
2. Changing this to a property changes the behavior: Previously, accessing the method without calling would not raise. Now, code that accesses the property (without applying the transform to coordinates) triggers the exception.

**Comment by stefanv** [comment]
> * Shouldn't it rather raise a specific exception class or `NotImplementedError`?

Yes!

> * Changing this to a property changes the behavior: Previously, accessing the method without calling would not raise. Now, code that accesses the property (without applying the transform to coordinates) triggers the exception.

Accessing the property should create the class. The class is callable by default, so you don't need to do anything else in addition.

File: skimage/transform/_geometric.py (hunk 4 above)
Comment (aeisenbarth) [reply skimage/transform/_geometric.py:288]: See the discussion above: https://github.com/scikit-image/scikit-image/pull/6926#discussion_r1190949714

`__call__` does `return coords_homogeneous @ self.params.T` and `inverse` did `return coords_homogeneous @ self.params`. So the new inverse transform instance should swap both.

File: skimage/transform/_geometric.py (hunk 7 above)
Comment (aeisenbarth) [reply skimage/transform/_geometric.py:1115]: I saw this proposed in the [issue](https://github.com/scikit-image/scikit-image/issues/6921#issuecomment-1542312019). I can also use explicit classes for the others as well, if preferred?

Comment (stefanv) [reply skimage/transform/_geometric.py:1115]: Not a big deal. I'd use `type` for general applications, and the specific class for the case where you know what the class name is up-front.

File: skimage/transform/tests/test_geometric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aeisenbarth) [reply skimage/transform/tests/test_geometric.py:571]: I added the suggested code.

**Comment by aeisenbarth** [comment]
I fully agree! But that is a bit tricky. If I move the exception message into the new transform's `__call__`, the transform must know that it is an inverse.

I [implemented](https://github.com/scikit-image/scikit-image/pull/6926/commits/9eb50c2a4fd1d4ae70e677a155ce85887855f9b5#diff-34acab621bfddc46a8c62f72a08284cd20d06058b76682476a2bc37a11f80b24R1620) an idea how this can be achieved. What do you think?

**Comment by stefanv** [comment]
> I fully agree! But that is a bit tricky. If I move the exception message into the new transform's `__call__`, the transform must know that it is an inverse.
> 
> I [implemented](https://github.com/scikit-image/scikit-image/pull/6926/commits/9eb50c2a4fd1d4ae70e677a155ce85887855f9b5#diff-34acab621bfddc46a8c62f72a08284cd20d06058b76682476a2bc37a11f80b24R1620) an idea how this can be achieved. What do you think?

Oh! Sorry, I think I misunderstood. I think it's fine to raise an exception on the spot where inverse is called on a piecewise affine transform. Sorry for the confusion!

**Comment by aeisenbarth** [comment]
You mean leaving the single `PolynomialTransform.inverse(coords)` as method and not converting it to a property? That is probably much more straight-forward than delegating the exception into a new transform instance, but it would make static analyzers unhappy. My IDE highlights that it mismatches the GeomTransform's interface, which is now a property and has no `coords` argument.

**Comment by stefanv** [comment]
We're dancing to the puppets of the machines again 😂

Something like: if inverse cannot be computed, raise exception, otherwise calculate the inverse now and create the class. That should be fine? Return class or raise.

File: skimage/transform/_geometric.py (hunk 7 above)
Comment (lagru) [reply skimage/transform/_geometric.py]: Actually, I think `type(self)` would be the better practice here. It will allow an inheriting classes X to use this property and return a new inverse of X (the same class). Do we want to explicitly block that behavior by using an explicit class here?

Comment (aeisenbarth) [reply skimage/transform/_geometric.py:1115]: That reasoning makes sense, I learned something! I reverted it back to `type(self)`.

**Comment by stefanv** [comment]
I think this is close to done. The inverse of the FundamentalMatrixTransform still bothers me. (a) It is excluded from the tests and (b) I am not sure it can be calculated as the transform---where did you see this?

**Comment by lagru** [comment]
> (b) I am not sure it can be calculated as the transform---where did you see this?

Not sure if I misunderstand you but https://github.com/scikit-image/scikit-image/pull/6926#discussion_r1190949714 points out a source why the transpose of the matrix corresponds to the fundamental matrix for the inverse transformation.

However, while trying to come up with a test to address your point (a), I noticed that applying the reverse transformation doesn't seem to work in all cases 
```python
import skimage as ski
SRC = np.array([
    [-12.3705, -10.5075],
    [-10.7865, 15.4305],
    [8.6985, 10.8675],
    [11.4975, -9.5715],
    [7.8435, 7.4835],
    [-5.3325, 6.5025],
    [6.7905, -6.3765],
    [-6.1695, -0.8235],
])
DST = np.array([
    [0, 0],
    [0, 5800],
    [4900, 5800],
    [4900, 0],
    [4479, 4580],
    [1176, 3660],
    [3754, 790],
    [1024, 1931],
])
tform1 = ski.transform.ProjectiveTransform(matrix=np.random.rand(3, 3))
tform2  = ski.transform.AffineTransform(scale=(0.1, 0.1), rotation=0.3)
tform3 = ski.transform.EuclideanTransform(rotation=0.9, translation=(5, 5))
tform4 = ski.transform.SimilarityTransform(scale=0.1, rotation=0.9)
for tform in (tform1, tform2, tform3, tform4):
    assert tform.estimate(SRC, DST)
    SRC_ = tform.inverse(tform(SRC))
    np.testing.assert_almost_equal(SRC, SRC_)
```
This passes. But for `EssentialMatrixTransform`, `FundamentalMatrixTransform` this fails with 
```
    ...
    return coords_homogeneous @ self.params
           ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 3 is different from 4)
```
:thinking:

File: skimage/transform/tests/test_geometric.py (hunk 2 above)
Comment (lagru) [reply skimage/transform/tests/test_geometric.py:562]: Huh? I thought I replied to this. I'm definitely either for adding a comment here or moving this bit into a helper function. It's a really cool and smart piece of code but probably not something that's easy to read for most people including me in a few months. :sweat_smile: 

I totally had to learn something new about how `and` works in Python! From https://docs.python.org/3/reference/expressions.html#boolean-operations

> Note that neither and nor or restrict the value and type they return to False and True, but rather return the last evaluated argument. This is sometimes useful, e.g., if s is a string that should be replaced by a default value if it is empty, the expression s or 'foo' yields the desired value. Because not has to create a new value, it returns a boolean value regardless of the type of its argument (for example, not 'foo' produces False rather than ''.)

**Comment by stefanv** [comment]
> Not sure if I misunderstand you but [#6926 (comment)](https://github.com/scikit-image/scikit-image/pull/6926#discussion_r1190949714) points out a source why the transpose of the matrix corresponds to the fundamental matrix for the inverse transformation.

Thanks, I missed that; we should add that as a reference.

**Comment by lagru** [comment]
It's already referenced here https://github.com/scikit-image/scikit-image/blob/5d254fc65ee623baecb13efac22fcc32032038a1/skimage/transform/_geometric.py#L247-L248

Concerning https://github.com/scikit-image/scikit-image/pull/6926#issuecomment-1549367974, looking through the tests it may be that I just can't really use SRC and DST for those transforms and need to adapt in- and output from their existing tests. Not sure whether I can look into this soon though...

**Comment by stefanv** [comment]
> It's already referenced here
> 
> https://github.com/scikit-image/scikit-image/blob/5d254fc65ee623baecb13efac22fcc32032038a1/skimage/transform/_geometric.py#L247-L248
> 
> Concerning [#6926 (comment)](https://github.com/scikit-image/scikit-image/pull/6926#issuecomment-1549367974), looking through the tests it may be that I just can't really use SRC and DST for those transforms and need to adapt in- and output from their existing tests. Not sure whether I can look into this soon though...

Because that code is confusing, we need at least a comment that explains that the transform is the inverse of the F matrix.

File: skimage/transform/tests/test_geometric.py [hunk 8]
```diff
@@ -354,6 +354,24 @@ def test_fundamental_matrix_inverse():
...
+    dst = np.array([1.002114, 1.129644, 1.521742, 1.846002,
+                    1.084332, 0.275134, 0.293328, 0.588992,
+                    0.839509, 0.087290, 1.779735, 1.116857,
+                    0.878616, 0.602447, 0.642616, 1.028681]).reshape(-1, 2)
+
+    tform = estimate_transform('fundamental', src, dst)
+
+    # calculate x' F x for each coordinate; should be close to zero
+    p = np.abs(np.sum(np.column_stack((dst, np.ones(len(dst)))) * tform(src), axis=1))
+    assert np.all(p < 0.01)
```
Comment (lagru) [inline skimage/transform/tests/test_geometric.py:372]: @stefanv could you maybe expand on this a bit more. I'm not really following. Your comment seems similar to Eq. 8.5 \$x'^T F x = 0\$, though there is a transpose missing? Also if it is indeed testing Eq. 8.5 how is this related to the property that \$F^T\$ describes the matrix for the reverse transformation?

Comment (stefanv) [reply skimage/transform/tests/test_geometric.py:372]: Yes, the test is named incorrectly! I started out wanting to test both directions, and ended up testing only one.

The convention in H&Z is x containing row-vectors, I think, which is the opposite of how we store them.

**Review by stefanv — APPROVED** [review]

**Comment by stefanv** [comment]
Thank you @aeisenbarth!

File: skimage/transform/tests/test_geometric.py [hunk 9]
```diff
@@ -354,6 +354,43 @@ def test_fundamental_matrix_inverse():
...
+                    1.084332, 0.275134, 0.293328, 0.588992,
+                    0.839509, 0.087290, 1.779735, 1.116857,
+                    0.878616, 0.602447, 0.642616, 1.028681]).reshape(-1, 2)
+
+    # Inverse of (src -> dst) transform should be equivalent to
+    # (dst -> src) transformation
+    tform = estimate_transform('fundamental', src, dst)
+    tform_inv = estimate_transform('fundamental', dst, src)
+
+    np.testing.assert_array_almost_equal(tform.inverse.params, tform_inv.params)
```
Comment (lagru) [inline skimage/transform/tests/test_geometric.py:373]: Nice. :) That test indeed increases my confidence in the correctness.
